### PR TITLE
🌱  Remove deprecated items from the codebase for v1beta1

### LINKED ISSUE #5254: Remove deprecated variables and functions for v1beta1 release
**Detailed Description**

There are a number of outstanding deprecation messages in the code that signal removal of a function, method or variable in the next version. As cluster API moves to release v1beta1 these items should be removed from the codebase in the main branch.

/kind feature

### PR DESCRIPTION
**Pull request by killianmuldoon**
Deprecated items across the code base are removed to make parsing
easier and reduce maintenance of the code going forward.

Signed-off-by: killianmuldoon <kmuldoon@vmware.com>

Fixes #5254

### CODE REVIEW

File: controllers/noderefutil/machine.go [hunk 1]
```diff
@@ -17,40 +17,9 @@ limitations under the License.
...
-			items = append(items, machine)
-		}
-	}
-	if len(items) != 1 {
-		return nil, errors.Errorf("expecting one machine for node %v, got %v", nodeName, machineNames(items))
-	}
-	return items[0], nil
-}
-
 func machineNames(machines []*clusterv1.Machine) []string {
```
Comment (killianmuldoon) [inline controllers/noderefutil/machine.go:23]: Needs to be deleted too!

**Review by killianmuldoon — COMMENTED** [review]
There are a few other deprecated types i.e. clusterStatus and the kubeadm bootstrap data which seem complex to remove and might get in the way of conversions working correctly from v1alpha3->v1beta1->v1alpha3. I've left those in place right now and added TODOs to bring them up for  discussion

File: api/v1alpha3/common_types.go [hunk 2]
```diff
@@ -123,50 +122,6 @@ type MachineAddresses []MachineAddress
 // In future versions, controller-tools@v2 might allow overriding the type and validation for embedded
 // types. When that happens, this hack should be revisited.
 type ObjectMeta struct {
```
Comment (sbueringer) [inline api/v1alpha3/common_types.go:124]: I don't think we can and should remove anything in our v1alpha3/v1alpha4 packages. It's a breaking change if we remove fields of our "frozen" old API versions

File: api/v1beta1/index/machine_test.go [hunk 3]
```diff
@@ -20,46 +20,12 @@ import (
...
 
 	. "github.com/onsi/gomega"
-	corev1 "k8s.io/api/core/v1"
 	"k8s.io/utils/pointer"
 	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
 	"sigs.k8s.io/cluster-api/controllers/noderefutil"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
-func TestIndexMachineByNodeName(t *testing.T) {
```
Comment (sbueringer) [inline api/v1beta1/index/machine_test.go:30]: Why was this test removed?

File: bootstrap/kubeadm/api/v1alpha3/kubeadmconfig_types.go [hunk 4]
```diff
@@ -109,7 +109,7 @@ type KubeadmConfigStatus struct {
 	// BootstrapData will be a cloud-init script for now.
 	//
 	// Deprecated: Switch to DataSecretName.
-	//
+	// TODO: This needs to be kept to support conversion to v1alpha3 types.
```
Comment (sbueringer) [inline bootstrap/kubeadm/api/v1alpha3/kubeadmconfig_types.go:112]: I'm not sure this TODO is helpful. 

We will drop the whole package at some point (when we don't support v1alpha3 anymore)

File: bootstrap/kubeadm/types/upstreamv1beta2/types.go [hunk 5]
```diff
@@ -126,6 +126,7 @@ type ClusterConfiguration struct {
 	//
 	// Deprecated: As hyperkube is itself deprecated, this fields is too. It will be removed in future kubeadm config versions, kubeadm
 	// will print multiple warnings when set to true, and at some point it may become ignored.
+	// TODO: Remove this field once conversions to Kubeadm v1beta2 --> v1alpha4 --> v1beta2 are no longer supported.
 	// +optional
```
Comment (sbueringer) [inline bootstrap/kubeadm/types/upstreamv1beta2/types.go:130]: I'm not sure this TODO is helpful. 

We will drop the whole package at some point (when we don't support the Kubernetes/kubeadm versions which require kubeadm v1beta2 types anymore)

Similar to this issue for v1beta1: https://github.com/kubernetes-sigs/cluster-api/issues/4507

File: cmd/clusterctl/api/v1alpha3/provider_type.go [hunk 6]
```diff
@@ -46,12 +46,6 @@ type Provider struct {
 	// Version indicates the component version.
 	// +optional
 	Version string `json:"version,omitempty"`
-
-	// WatchedNamespace indicates the namespace where the provider controller is is watching.
-	// if empty the provider controller is watching for objects in all namespaces.
-	// Deprecated: in clusterctl v1alpha4 all the providers watch all the namespaces; this field will be removed in a future version of this API
-	// +optional
-	WatchedNamespace string `json:"watchedNamespace,omitempty"`
```
Comment (sbueringer) [inline cmd/clusterctl/api/v1alpha3/provider_type.go:54]: We cannot drop this field. As mentioned above, in my opinion we cannot ever remove fields in released API versions, we can drop them in future versions. (in this case there is no newer version than v1alpha3 for this type)

File: cmd/clusterctl/client/cluster/cert_manager.go [hunk 7]
```diff
@@ -44,7 +44,8 @@ const (
 
 	// Deprecated: Use clusterctlv1.CertManagerVersionAnnotation instead.
 	// This is maintained only for supporting upgrades from cluster created with clusterctl v1alpha3.
-	certManagerVersionAnnotation = "certmanager.clusterctl.cluster.x-k8s.io/version"
+	// TODO: Remove once upgrades from v1alpha3 are no longer supported.
+	certmanagerVersionAnnotation = "certmanager.clusterctl.cluster.x-k8s.io/version"
```
Comment (sbueringer) [inline cmd/clusterctl/client/cluster/cert_manager.go:48]:
```suggestion
	certManagerVersionAnnotation = "certmanager.clusterctl.cluster.x-k8s.io/version"
```
I don't think we should change the var name

File: controllers/mdutil/util.go [hunk 8]
```diff
@@ -42,27 +42,6 @@ const (
 	// DefaultMachineDeploymentUniqueLabelKey is the label applied to Machines
 	// in a MachineDeployment containing the hash of the template.
 	DefaultMachineDeploymentUniqueLabelKey = "machine-template-hash"
```
Comment (sbueringer) [inline controllers/mdutil/util.go:44]: Note/fyi This whole package will be deprecated and removed in a follow-up PR to https://github.com/kubernetes-sigs/cluster-api/pull/5250/files

(but your change here is fine)

**Review by sbueringer — COMMENTED** [review]
Can you please add the removed fields/funcs/constants/... to the v1alpha4=>v1beta1 migration doc, similar to what we did in https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/book/src/developer/providers/v1alpha3-to-v1alpha4.md (possibly in a more compact format)

File: cmd/clusterctl/api/v1alpha3/provider_type.go (hunk 6 above)
Comment (killianmuldoon) [reply cmd/clusterctl/api/v1alpha3/provider_type.go:54]: Thanks for pointing that out - I'll revise the PR to keep everything under v1alpha3/4 types the same and only remove types from v1beta1

File: bootstrap/kubeadm/types/upstreamv1beta2/types.go (hunk 5 above)
Comment (killianmuldoon) [reply bootstrap/kubeadm/types/upstreamv1beta2/types.go:130]: Agreed - I'll revert the kubeadm types too

File: cmd/clusterctl/client/cluster/cert_manager.go (hunk 7 above)
Comment (killianmuldoon) [reply cmd/clusterctl/client/cluster/cert_manager.go:48]: typo - will revert.

File: api/v1beta1/index/machine_test.go (hunk 3 above)
Comment (killianmuldoon) [reply api/v1beta1/index/machine_test.go]: In error - I've replaced it now.

File: cmd/clusterctl/client/cluster/cert_manager.go [hunk 9]
```diff
@@ -44,6 +44,7 @@ const (
 
 	// Deprecated: Use clusterctlv1.CertManagerVersionAnnotation instead.
 	// This is maintained only for supporting upgrades from cluster created with clusterctl v1alpha3.
+	// TODO: Remove once upgrades from v1alpha3 are no longer supported.
```
Comment (enxebre) [inline cmd/clusterctl/client/cluster/cert_manager.go:47]: fwiw this is unexported anyway so won't make a difference for consumers.

File: cmd/clusterctl/client/upgrade.go [hunk 10]
```diff
@@ -208,9 +208,6 @@ func parseUpgradeItem(ref string, providerType clusterctlv1.ProviderType) (*clus
 			},
 			ProviderName: name,
 			Type:         string(providerType),
-			// The value for the following fields will be retrieved while
-			// creating the custom upgrade plan.
-			WatchedNamespace: "",
```
Comment (enxebre) [inline cmd/clusterctl/client/upgrade.go:213]: As per https://github.com/kubernetes-sigs/cluster-api/pull/5288/files#r713637308 this behaviour can't be broken unless the api is bumped.

File: cmd/clusterctl/api/v1alpha3/provider_type.go (hunk 6 above)
Comment (sbueringer) [reply cmd/clusterctl/api/v1alpha3/provider_type.go:54]: @killianmuldoon I think we have to keep this one (not sure if it's an oversight or it's WIP to keep it)

File: cmd/clusterctl/client/upgrade.go (hunk 10 above)
Comment (sbueringer) [reply cmd/clusterctl/client/upgrade.go:213]: +1 Imho we have to keep the field, see https://github.com/kubernetes-sigs/cluster-api/pull/5288#discussion_r713637308

File: docs/book/src/developer/providers/v1alpha4-to-v1beta1.md [hunk 11]
```diff
@@ -21,3 +21,32 @@
...
+
+## Removed items
+
+### API Fields 
+
+- **ClusterTopologyLabelName**, a ClusterClass related constant has been deprecated and removed. This label has been replaced by `ClusterTopologyOwnedLabel`.
+
+- **MachineNodeNameIndex** has been removed from the common types in favor of `api/v1beta1/index.MachineNodeNameField`.
+
+- **MachineProviderNameIndex** has been removed from common types in favor of `v1beta1/index.MachineProviderIDField`.
```
Comment (sbueringer) [inline docs/book/src/developer/providers/v1alpha4-to-v1beta1.md:33]:
```suggestion
- **MachineProviderNameIndex** has been removed from common types in favor of ``api/v1beta1/index.MachineProviderIDField`.
```
nit

**Review by sbueringer — COMMENTED** [review]
last nit from my side. Otherwise looks great!

**Review by vincepri — APPROVED** [review]
/approve

**Pull request merged**